pyswarms/backend/handlers.py

```diff
@@ -540,3 +542,193 @@ def zero(self, velocity, clamp=None, **kwargs):
             raise
         else:
             return new_vel
+
+
+class OptionsHandler(HandlerMixin):
+    def __init__(self, strategy):
+        """An OptionsHandler class
+
+        This class offers a way to handle options. It contains
+        methods to vary the options at runtime.
+        Following strategies are available for the handling:
+
+        * exp_decay:
+            Decreases the parameter exponentially between limits.
+
+        * lin_decay:
+            Decreases the parameter linearly between limits.
+
+        * random:
+            takes a uniform random value between (0.5,1)
+
+        * nonlin_mod:
+            Decreases the parameter between limits according to a nonlinear modulation index .
+
+        The OptionsHandler can be called as a function to use the strategy
+        that is passed at initialization to account for time-varying coefficients. An example
+        for the usage:
+
+        .. code-block :: python
+
+            from pyswarms.backend import operators as op
+            from pyswarms.backend.handlers import OptionsHandler
+
+
+            oh = OptionsHandler(strategy=(None, None, "exp_decay"))
```

Comment:
Are we expecting the user to use multiple strategies at the same time? Or do we just expect a single one every time?


1. If multiple, then maybe passing a list, without explicitly stating `None`s should be better:

```python
# If expecting multiple strats
oh = OptionsHandler(strategy=["exp_decay", "lin_decay"])  # will just apply exp_decay and lin_decay

# If single strat only
oh = OptionsHandler(strategy=["exp_decay"])
```

Then within your code you can just iterate like so:

```python
# Assuming: strategies =  {"exp_decay": exp_decay, "lin_decay": lin_decay, "some_decay": some_decay}
choices = ["exp_decay", "lin_decay"]

for choice in choices:
    strategy = strategies[choice]
    strategy()
```

2. If we're just expecting a single strategy from the user every time, then I think passing a string is way better

```python
# If single strat only
oh = OptionsHandler(strategy=["exp_decay"])
```

Reply:
Hi @ljvmiranda921! Thanks for the review. Would love to contribute more.

We definitely want to allow different strategies for different options. This might be useful in something like a Time-varying acceleration PSO. 
**Why use a tuple not a list:**
1(a) A list could surely be used without any change but I purposely used a tuple as part of an attempt to avoid mutable arguments. Although the strategy iterable is not getting mutated within any function, it was just to avoid any confusion and is just good python practice.

**Why use an explicit None**
1(b) Say the user wants to do the following: 
Nonlinear modulation decay for the cognitive component (c1)
Nothing(constant) for the social component (c2)
Exponential decay for the inertia weight(w)

```python3
options = {"c1": 0.5, "c2": 0.3, "w": 0.9}   # will be used as the start options 

# A if we dont have explicit None`
oh_strategy=("nonlin_mod", "exp_decay")

# B if we have explicit None
oh_strategy=("nonlin_mod", None,  "exp_decay")

# Call instance of PSO
optimizer = ps.single.GeneralOptimizerPSO(
    n_particles=10,
    dimensions=2,
    topology=Star(),
    options=options,
    oh_strategy=oh_strategy,
)
cost, pos = optimizer.optimize(fx.sphere, iters=100)

# inside the call of options handler it is something similar to 
for key, strat in zip(options.keys(), oh_strategy):
    if strat is None:
        continue
    "change the option with "key" to value corresponding to "strat"
```
Since we have to loop over the strategy tuple, the first option will take up the first strategy and so forth.
This means that with method A we will result in an exponential decay for the second option i.e. "c2". Which is wrong because we need it to be constant.
This need for correct order is definitely a small weakness in some sense but the alternative solution was to use a dict (or a namedTuple maybe?) for the oh_strategy which wasn't very consistent with the other handlers and is also mutable again.

2. Agreed. If the same option for all coefficients was better in terms of functionality, a string would be best.

Let me know what you think! I dont mind changing anything and everything as long as its better in terms of style, convention, speed etc.

Reply:
> Since we have to loop over the strategy tuple, the first option will take up the first strategy and so forth.
This means that with method A we will result in an exponential decay for the second option i.e. "c2". Which is wrong because we need it to be constant.

Hi @nikhil-sethi now makes sense why you opted for this data structure! Because it applies to c1, c2, and w. Hmmm, perhaps it's fine to make it explicit rather than implying based on index positions (first position is c1, second position is c2, third position is w). This then allows more readability on the user side and reduces any surprises in the long run! 

 > This need for correct order is definitely a small weakness in some sense but the alternative solution was to use a dict (or a namedTuple maybe?) for the oh_strategy which wasn't very consistent with the other handlers and is also mutable again.

I'm fine using a dictionary for this, even if it's not consistent with the other handlers. I guess you just need to override some of the inherited methods so that it accepts your data structure. How about this one?

```python
options = {"c1": 0.5, "c2": 0.3, "w": 0.9}   # will be used as the start options 
# Multiple options
oh_strategy={"w":"exp_decay", "c1":"nonlin_mod"}
# Single option
oh_strategy={"c1":"nonlin_mod"}
``` 

Let me know what you think!

Reply:
Okay. I agree it's more readable and intuitive. I didn't want to do it before without discussing it as it would have broken style.  I'll make changes as soon as I can :)

Reply:
@nikhil-sethi Looking forward to it!